chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,467 @@
+        self.max_n_clusters = max_n_clusters
+
+        # L1 or L2 norm is allowed only
+        if norm == 1:
+            self.norm = self._l1_norm
+        elif norm == 2:
+            self.norm = self._l2_norm
+        else:
+            raise ValueError("Norm can only be 1 or 2")
+
+    def _l1_norm(self, x, axis):
+        """Function to calculate L1 Norm by given axes
+
+        Args:
+            x (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`):
+                Input matrix to calculate norm
+            axis (int / tuple): Axes information for norm calculation
+
+        Returns:
+            :class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray` : Norm applied to given axes
+        """
+        return c_sum(absolute(x), axis=axis)
+
+    def _l2_norm(self, x, axis):
+        """Function to calculate L2 Norm by given axes
+
+        Args:
+            x (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`):
+                Input matrix to calculate norm
+            axis (int / tuple): Axes information for norm calculation
+
+        Returns:
+            :class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray` :
+                Norm applied to given axes
+        """
+        return c_sum(x ** 2, axis=axis)
+
+    def _variance_term(self, pred, gt, means, delta_v, gt_idx):
+        """Function to calculate variance term
+
+        Args:
+            pred (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or '
+            :class:`cupy.ndarray`):
+                Prediction output
+            gt (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or '
+            :class:`cupy.ndarray`):
+                Ground truth output
+            means (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or '
+            :class:`cupy.ndarray`):
+                Instance means
+            delta_v (float): Coefficient to decide 'pull force' power
+            gt_idx (tuple / nd-array): Indexes of ground truth instances
+
+        Returns:
+            :class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray` : variance loss
+        """
+
+        bs, n_filters, n_loc = pred.shape
+        n_instances = gt.shape[1]
+
+        # Prepare each item with same size by broadcasting missing axes
+        p = expand_dims(pred, axis=1)
+        g = expand_dims(gt, axis=2)
+        m = expand_dims(means, axis=3)
+
+        p = broadcast_to(p, (bs, n_instances, n_filters, n_loc))
+        g = broadcast_to(g, (bs, n_instances, n_filters, n_loc))
+        m = broadcast_to(m, (bs, n_instances, n_filters, n_loc))
+
+        m = cast(m, p.dtype)
+        g = cast(g, p.dtype)
+
+        module = cuda.get_array_module(p)
```

Comment:
Please name `module` as `xp` instead, since this is the convention in other Chainer code, and makes the function easier to read.